### ISSUE TITLE
Using ACF block.json fields in a native Gutenberg block

### ISSUE THREAD

**Issue by dansdev** [AdvancedCustomFields/acf#727]
Hi there,

With the latest update of ACF (6.0) ACF Blocks can now be registered the recommended WP way of doing things using block.json. Is it possible to register ACF block.json and use Local JSON fields within a @wordpress/create-block plugin?

By using Local JSON, I'm taking a guess that ACF would need to be included in the plugin?

Reason being, @wordpress/create-block wpscripts comes with really helpful Webpack build tooling. It is also now possible to register multiple blocks within a @wordpress/create-block plugin, which would negate the need to setup a package.json and webpack.config etc..for each registered ACF block.

I'll be interested to know whether this is possible or maybe there is another (better) way of achieving this?

**Comment by colorful-tones** [comment]
This may be of interest: https://github.com/WordPress/gutenberg/issues/44626

**Comment by colorful-tones** [comment]
@lgladdy I've created an issue ☝️ for the `@wordpress/create-block` package to try and get some traction around adding the ability to extend Create Block Template's and allow custom keys to be passed. This would help ACF block builders a good deal I believe. I just wanted to put it on the radar over in this repo.

**Comment by colorful-tones** [comment]
From the ACF side of things. It would be awesome to be able to get an additional field in ACF's key to pass through the local JSON file for ACF fields and pull them in. Something like:

```
{
    "name": "acf-example-block",
    "title": "ACF Example Block",
    "description": "ACF example block",
    "apiVersion": 2,
    "acf": {
        "mode": "preview"
        "localJSON": "file:./acf-example-block-fields.json"
    }
}
```

**Comment by dansdev** [comment]
Hey @colorful-tones,

Thanks for the interest and creating the issue for the `@wordpress/create-block` package. I agree the additional ACF key to pass through the local JSON would be amazing! Especially from within a plugin, would negate the need to include the ACF core files to the plugin directory. 

I had bit more of a play around and managed to get ACF blocks field rendering through the `@wordpress/create-block` block.json within a theme. I had to configure the initial `@wordpress/create-block` package output, but it does work! For anyone else who's trying to get this working within a theme this is how I had things structured:

```
theme
- blocks (this is where you install @wordpress/create-block)
-- build
-- package.json
-- init.php (blocks are registered here, required this file from functions.php)
-- src (additional blocks can be added here after the initial @wordpress/create-block install using the --no-plugin flag)
--- acf-block (block folder)
---- block.json
---- index.js (import .scss file here)
---- render.php (this contains the block output acf fields etc..)
---- style.scss
```

```
block.json
{
	"apiVersion": 2,
	"name": "create-block/acf-block",
	"version": "0.1.0",
	"title": "ACF Block",
	"category": "widgets",
	"icon": "smiley",
	"description": "ACF Block",
	"acf": {
            "mode": "preview",
            "renderTemplate": "render.php"
          },
	"supports": {
          "align": false,
          "anchor": false,
          "alignContent": false,
          "color": {},
          "alignText": false,
          "spacing": {},
          "typography": {},
          "fullHeight": false
	},
	"textdomain": "acf-block",
	"script": "file:./index.js",
	"style": "file:./style-index.css"
}
```

```
init.php (A for each file in directory function would be nice here, to save manually registering each block)

function theme_blocks_block_init() {
	register_block_type( get_stylesheet_directory() . '/blocks/build/acf-block' );
}
add_action( 'init', 'theme_blocks_block_init' );
```

As the I'm using the ACF renderTemplate, the render.php needs to copied with the src files. wp-scripts has the webpack php copy plugin included with it. So this build script in the package.json can be used `"build:copy-php": "wp-scripts build --webpack-copy-php",`.

**Comment by colorful-tones** [comment]
> I agree, having the additional ACF key to pass through the local JSON would be amazing! Would that mean that the core ACF files wouldn’t need to be included in the plugin directory?

No, ACF would still need to be activated on the site as all the fields would not be operable without it.

> I had bit more of a play around and managed to get ACF block fields rendering through the @wordpress/create-block block.json within a theme. I had to configure the initial @wordpress/create-block package output, but it does work!

Yes, you can still use `@wordpress/create-block` to generate things, but it takes a bit of manual work right now. I'm trying to allow us to be able to generated a custom Create Block Template, which we could publish and then folks could either run it in a theme or plugin. Since `@wordpress/create-block` allows the creation of plugin too, which is kind of ideal. I recommend blocks be stored in a plugin. This way, if the theme is changed then the blocks should still be available to the editor.

There is already progress being made on: https://github.com/WordPress/gutenberg/issues/44626

In my ideal scenario, I would like to be able to run `npx @wordpress/create-block my-cool-block --template=./block-template/index.js --no-plugin --no-wp-scripts --variant=dynamic`, which would output:

`my-cool-block/block.json`

```
{
	"$schema": "https://schemas.wp.org/trunk/block.json",
	"apiVersion": 2,
	"name": "neat-namespace/my-cool-block",
	"version": "0.1.0",
	"title": "My cool block",
	"category": "text",
	"description": "A pretty cool block",
	"textdomain": "neat-namespace",
	"editorStyle": "file:./editor.css",
	"style": "file:./frontend.css",
        "viewScript": "file:./frontend.js",
	"acf": {
		"mode": "preview"
	},
	"render": "file:./render.php",
}
```

As well as:

```
my-cool-block/
    render.php
    editor.css
    frontend.css
    frontend.js
```

**Comment by dansdev** [comment]
Hey @colorful-tones,

Thanks for info, it looks you guys are making great progress on https://github.com/WordPress/gutenberg/issues/44626!

Quick question on your block.json example if you have the time - When the new ACF key has been added, will the render property replace the need for renderTemplate used in ACF?

> ```
> {
> 	"$schema": "https://schemas.wp.org/trunk/block.json",
> 	"apiVersion": 2,
> 	"name": "neat-namespace/my-cool-block",
> 	"version": "0.1.0",
> 	"title": "My cool block",
> 	"category": "text",
> 	"description": "A pretty cool block",
> 	"textdomain": "neat-namespace",
> 	"editorStyle": "file:./editor.css",
> 	"style": "file:./frontend.css",
>         "viewScript": "file:./frontend.js",
> 	"acf": {
> 		"mode": "preview"
> 	},
> 	"render": "file:./render.php",
> }
> ```

```
    "acf": {
        "mode": "preview",
        "renderTemplate": "render.php"
    },
```

Cheers!

**Comment by colorful-tones** [comment]
> will the render property replace the need for renderTemplate used in ACF?

The `render` will [be in WP 6.1](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#render) for native blocks. I'm not sure how this may intersect with ACF. 🤔

**Comment by chrillep** [comment]
this schema contains acf aswell https://gist.github.com/saulirajala/359775273c62fa19ea232ced18c10712
would be nice if acf would maintain it instead
